SmartTutor/ContentModel/LearningObjects/Challenges/FulfillmentStrategy/ProjectChecker/ProjectAnalyzer.cs

```diff
@@ -0,0 +1,70 @@
+﻿using CodeModel.CaDETModel.CodeItems;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SmartTutor.ContentModel.LearningObjects.Challenges.FulfillmentStrategy.ProjectChecker
+{
+    public class ProjectAnalyzer : ChallengeFulfillmentStrategy
+    {
+        public Dictionary<string, List<ChallengeFulfillmentStrategy>> SnippetApplicableStrategies { get; private set; }
+
+        private ProjectAnalyzer() { }
+
+        public ProjectAnalyzer(Dictionary<string, List<ChallengeFulfillmentStrategy>> snippetApplicableStrategies) : this()
+        {
+            SnippetApplicableStrategies = snippetApplicableStrategies;
+        }
+
+        public override HintDirectory EvaluateSubmission(List<CaDETClass> solutionAttempt)
+        {
+            var challengeHints = new HintDirectory();
+            foreach (var strategy in SnippetApplicableStrategies.SelectMany(sas => sas.Value))
+            {
+                var result = strategy.EvaluateSubmission(GetAllClassesFromSnippets(solutionAttempt));
```

Comment:
If I'm not mistaken, this code takes all the strategies and applies them to all the CaDETClasses, asking each one if it contains the suitable snippet below. This has a major readability issue - the EvaluateSubmission is a high-level method that is read first and that should give us an idea of what the class is doing without requiring us to read the details. When I'm reading this method I see that all the strategies from the dictionary are applied to all the classes (from snippets?).

Only when I go into the details do I see that the key is used (and the below functions are confusing to grasp - at least more than they should be).

A more straightforward flow of logic would be:

- For each code snippet for which we have strategies:
    - Find the code item (either CaDETClass or CaDETMember) based on the snippet id. Throw an exception if there is no such item.
    - Apply the strategies related to that snippet.
    - Merge the obtained hints.

Changing this to code, we would have something along the lines of:


    foreach (var codeSnippetId in SnippetApplicableStrategies.Keys) {
        var codeSnippet = FindCodeSnippet(codeSnippetId, solutionAttempt); //if snippedId contains () looks at method or constructor signatures and returns empty class with that member, otherwise classes. Throws exception if not found.
        foreach(var strategy in SnippetApplicableStrategies[codeSnippetId]) {
            challengeHints.MergeHints(strategy.EvaluateSubmission(codeSnippet);
        }
    }